Rationalizing floats in the intersection function in util.py

codecruisedor opened the pull request:
Added functionality to rationalize floats in the intersection function in util.py

**References to other Issues or PRs**
Fixes [#25368](https://github.com/sympy/sympy/issues/25368)

**Brief description of what is fixed or changed**
Geometric Entities that have float values in them are rationalized in the intersection finding method in `util.py`. This intends to solve the problem of not finding intersections when having entities with float values.

**Release Notes**
<!-- BEGIN RELEASE NOTES -->
* geometry
  * Computes the correct intersection result when using floats.
<!-- END RELEASE NOTES -->

P.S. This is my first pull request in this project, please feel free to point out any mistake or optimization that you would like me to fix.

codecruisedor committed:
Commit-1  for issue #25368

Added a rationalizing method in the util.py file that is used by the intersection method.

anutosh491 commented:
Hello @codecruisedor . I would recommend going through this link https://github.com/sympy/sympy/wiki/Writing-Release-Notes on framing Release Notes and add one instead of **NO ENTRY**

anutosh491 commented:
I like the discussion you had with @smichr on the issue and that looks like the correct way to proceed. You could let us know once you are able to make the tests pass.

codecruisedor commented:
Sure, @anutosh491. Thank you very much for the help.

smichr commented:
You do not need a new function. nsimplify(ellipse, rational=True) will work.

codecruisedor commented:
Most entities except `ellipse` seem to rationalize floats internally. Should I check the type of entity before using `nsimplify` on it..? Or use it on every entity regardless? 

Like this (inside the intersection function):

```
    y = Dummy()
    for i, e in enumerate(entities):
        if not isinstance(e, GeometryEntity):
            entities[i] = Point(e)
        entities[i] = entities[i].replace(lambda x: x.is_Float, lambda x: nsimplify(x, rational=True) + y).xreplace({y: 0})
```

codecruisedor committed:
Merge remote-tracking branch 'upstream/master' into codecruisedor-rationalize_floats_in_intersection

codecruisedor committed:
Commit-2 : Integrated the functionality inside intersection method

Instead of a separate function, the change is done in the intersection function itself.

codecruisedor committed:
Commit -3 : Removed unused import

Removed unused import from test.util

smichr commented on sympy/geometry/tests/test_util.py:
```diff
@@ -57,6 +57,8 @@ def test_intersection():
         Point(0, 0), Segment((0, 0), (1, 0))]
 
 
+
+
```
```suggestion
```

smichr replied:
We need a test where the intersection that failed in master will now give the correct result. Also, if the entity was given in terms of Floats, the intersection should probably be returned in Float, too, with a precision matching the highest precision number in the expression.

smichr replied:
This is waiting for a test.

codecruisedor replied:
Yes, working on it

smichr replied:
I just applied a change that tracks the precision -- you might need to pull this locally and then add the test. If you `git push -f codecruisedor` you will wipe out the changes I made. (And feel free to do that.) In case you do and want to see the changes again, they are:
```python
def intersection(*entities, pairwise=False, **kwargs):
    """The intersection of a collection of GeometryEntity instances.

    Parameters
    ==========
    entities : sequence of GeometryEntity
    pairwise (keyword argument) : Can be either True or False

    Returns
    =======
    intersection : list of GeometryEntity

    Raises
    ======
    NotImplementedError
        When unable to calculate intersection.

    Notes
    =====
    The intersection of any geometrical entity with itself should return
    a list with one item: the entity in question.
    An intersection requires two or more entities. If only a single
    entity is given then the function will return an empty list.
    It is possible for `intersection` to miss intersections that one
    knows exists because the required quantities were not fully
    simplified internally.
    Reals should be converted to Rationals, e.g. Rational(str(real_num))
    or else failures due to floating point issues may result.

    Case 1: When the keyword argument 'pairwise' is False (default value):
    In this case, the function returns a list of intersections common to
    all entities.

    Case 2: When the keyword argument 'pairwise' is True:
    In this case, the functions returns a list intersections that occur
    between any pair of entities.

    See Also
    ========

    sympy.geometry.entity.GeometryEntity.intersection

    Examples
    ========

    >>> from sympy import Ray, Circle, intersection
    >>> c = Circle((0, 1), 1)
    >>> intersection(c, c.center)
    []
    >>> right = Ray((0, 0), (1, 0))
    >>> up = Ray((0, 0), (0, 1))
    >>> intersection(c, right, up)
    [Point2D(0, 0)]
    >>> intersection(c, right, up, pairwise=True)
    [Point2D(0, 0), Point2D(0, 2)]
    >>> left = Ray((1, 0), (0, 0))
    >>> intersection(right, left)
    [Segment2D(Point2D(0, 0), Point2D(1, 0))]

    """
    if len(entities) <= 1:
        return []

    entities = list(entities)
    prec = None
    for i, e in enumerate(entities):
        if not isinstance(e, GeometryEntity):
            # entities may be an immutable tuple
            e = Point(e)
        # convert to exact Rationals
        d = {}
        for f in e.atoms(Float):
            prec = f._prec if prec is None else min(f._prec, prec)
            d.setdefault(f, nsimplify(f, rational=True))
        entities[i] = e.xreplace(d)

    if not pairwise:
        # find the intersection common to all objects
        res = entities[0].intersection(entities[1])
        for entity in entities[2:]:
            newres = []
            for x in res:
                newres.extend(x.intersection(entity))
            res = newres
    else:
        # find all pairwise intersections
        ans = []
        for j in range(len(entities)):
            for k in range(j + 1, len(entities)):
                ans.extend(intersection(entities[j], entities[k]))
        res = list(ordered(set(ans)))
    # convert back to Floats
    if prec is not None:
        p = prec_to_dps(prec)
        res = [i.n(p) for i in res]
    return res
```

codecruisedor replied:
Okay, thanks👍..I'll add the test and commit.

smichr commented on sympy/geometry/util.py:
```diff
@@ -696,9 +697,11 @@ def intersection(*entities, pairwise=False, **kwargs):
     for i, e in enumerate(entities):
         if not isinstance(e, GeometryEntity):
             entities[i] = Point(e)
+        entities[i] = entities[i].replace(lambda x: x.is_Float, lambda x: nsimplify(x, rational=True) + y).xreplace({y: 0})
```
```suggestion
        entities[i] = entities[i].replace(lambda x: x.is_Float, lambda x: nsimplify(x, rational=True))
```

smichr replied:
I am not sure why you are doing this. What is `y` for?

codecruisedor replied:
I am not entirely certain as to why, but using the dummy object `y` gives out the correct result when using floats in ellipsis. Removing `y` results in an empty intersection result from what I have tested(as if the rationalization did not take effect!).

smichr replied:
My guess is that this is needed so the replacement will take place since otherwise 2 would not replace 2.0

smichr committed:
track smallest prec and return intersections accordingly

smichr committed:
import Float

codecruisedor commented:
Hi @smichr, just for clarification before I commit, here is the test I am writing for this(it is the test case that was provided in the original issue):

`    assert intersection(
            Ray(Point2D(0.001, -1),
            Point2D(0.0008, -1.7)),
            Ellipse(center=Point2D(0, 0), hradius=4.0, vradius=2.0), pairwise=True) == [
        Point(0.000714285723396502, -1.99999996811224, evaluate=False)]`

Will this suffice or do we need to add any other test..?

smichr commented:
If that passes, then I would just add another where the precision of one of the floats is set small, e.g. change 4.0 to Float(4, 10) to see that lower precision values are given in the Point.

codecruisedor commented:
Hi @smichr , the tests pass with `pytest.approx`. 

```
    assert intersection(
            Ray(Point2D(0.001, -1),
            Point2D(0.0008, -1.7)),
            Ellipse(center=Point2D(0, 0), hradius=4.0, vradius=2.0), pairwise=True)[0].coordinates == pytest.approx(
            Point2D(0.000714285723396502, -1.99999996811224, evaluate=False).coordinates
    )
    assert intersection(
            Ray(Point2D(0.001, -1),
            Point2D(0.0008, -1.7)),
            Ellipse(center=Point2D(0, 0), hradius=Float(4,10), vradius=2.0), pairwise=True)[0].coordinates == pytest.approx(
            Point2D(0.000714285723396502, -1.99999996811224, evaluate=False).coordinates
    )
```
have added a test with `Float` class. Any comments..?

smichr commented:
So add this is the area where intersections are tested, commit and push here.

codecruisedor committed:
Added tests in test_intersection

Added two assertion tests in the 'test_intersection' function

smichr commented on sympy/geometry/tests/test_util.py:
```diff
@@ -55,6 +57,16 @@ def test_intersection():
             Segment((-1, 0), (1, 0)),
             Line((0, 0), slope=1), pairwise=True) == [
         Point(0, 0), Segment((0, 0), (1, 0))]
+    assert intersection(
+            Ray(Point2D(0.001, -1),
+            Point2D(0.0008, -1.7)),
+            Ellipse(center=Point2D(0, 0), hradius=4.0, vradius=2.0), pairwise=True)[0].coordinates == pytest.approx(
+            Point2D(0.000714285723396502, -1.99999996811224, evaluate=False).coordinates)
+    assert intersection(
+            Ray(Point2D(0.001, -1),
+            Point2D(0.0008, -1.7)),
+            Ellipse(center=Point2D(0, 0), hradius=Float(4,10), vradius=2.0), pairwise=True)[0].coordinates == pytest.approx(
+            Point2D(0.000714285723396502, -1.99999996811224, evaluate=False).coordinates)
```
```suggestion
    R = 4.0
    intersection(
            Ray(Point2D(0.001, -1),
            Point2D(0.0008, -1.7)),
            Ellipse(center=Point2D(0, 0), hradius=R, vradius=2.0), pairwise=True)[0].coordinates
    assert c == pytest.approx(
            Point2D(0.000714285723396502, -1.99999996811224, evaluate=False).coordinates)
    # check this is responds to a lower precision parameter
    R = Float(4, 5)
    c2 = intersection(
            Ray(Point2D(0.001, -1),
            Point2D(0.0008, -1.7)),
            Ellipse(center=Point2D(0, 0), hradius=R, vradius=2.0), pairwise=True)[0].coordinates
    assert c2 == pytest.approx(
            Point2D(0.000714285723396502, -1.99999996811224, evaluate=False).coordinates)
    assert c[0]._prec == 53
    assert c2[0]._prec == 20
```

smichr committed:
show prec in test

smichr commented:
I just modified test to show that the coordinate precision not only passed but that it also has a different precision

smichr commented on sympy/geometry/tests/test_util.py:
```diff
@@ -55,6 +57,23 @@ def test_intersection():
             Line((0, 0), slope=1), pairwise=True) == [
         Point(0, 0), Segment((0, 0), (1, 0))]
+    R = 4.0
+    intersection(
```
```suggestion
    c = intersection(
```

smichr committed:
Update sympy/geometry/tests/test_util.py

smichr committed:
Update test_ellipse.py

smichr commented on sympy/geometry/tests/test_ellipse.py:
```diff
@@ -254,7 +254,7 @@ def line_list_close(ll1, ll2, prec):
     assert intersection(Circle(Point(0, 0), 2), Circle(Point(3, 0), 1)) == [Point(2, 0)]
     assert intersection(Circle(Point(0, 0), 2), Circle(Point(7, 0), 1)) == []
-    assert intersection(Ellipse(Point(0, 0), 5, 17), Ellipse(Point(4, 0), 1, 0.2)) == [Point(5, 0)]
+    assert intersection(Ellipse(Point(0, 0), 5, 17), Ellipse(Point(4, 0), 1, 0.2)) == [Point(5.0, 0)]
```
```suggestion
    assert intersection(Ellipse(Point(0, 0), 5, 17), Ellipse(Point(4, 0), 1, 0.2)
        ) == [Point(5.0, 0, evaluate=False)]
```

smichr committed:
Update sympy/geometry/tests/test_ellipse.py
